### PR TITLE
VMware: Add module vmware_vm_storage_policy_info

### PR DESCRIPTION
**Pull request by Akasurde**
##### SUMMARY
Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
lib/ansible/module_utils/vmware.py
lib/ansible/module_utils/vmware_spbm.py
lib/ansible/modules/cloud/vmware/vmware_vm_storage_policy_info.py
test/integration/targets/vmware_vm_storage_policy_info/aliases
test/integration/targets/vmware_vm_storage_policy_info/tasks/main.yml

### CODE REVIEW

File: lib/ansible/modules/cloud/vmware/vmware_vm_storage_policy_info.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+ANSIBLE_METADATA = {
+    'metadata_version': '1.1',
+    'status': ['preview'],
+    'supported_by': 'community'
+}
+
+DOCUMENTATION = r'''
+---
+module: vmware_vm_storage_policy_info
+short_description: Gather information about vSphere storage profile defines storage policy information.
```
Comment (jillr) [inline lib/ansible/modules/cloud/vmware/vmware_vm_storage_policy_info.py]:
```suggestion
short_description: Gather information about vSphere storage profile defined storage policy information.
```

File: lib/ansible/modules/cloud/vmware/vmware_vm_storage_policy_info.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+description:
+- Returns basic facts vSphere storage profiles.
```
Comment (jillr) [inline lib/ansible/modules/cloud/vmware/vmware_vm_storage_policy_info.py]:
```suggestion
- Returns basic facts about vSphere storage profiles.
```

**Review by jillr — COMMENTED** [review]
Couple of docs nits, otherwise lgtm

File: lib/ansible/module_utils/vmware.py [hunk 3]
```diff
@@ -490,7 +490,7 @@ def vmware_argument_spec():
     )
 
 
-def connect_to_api(module, disconnect_atexit=True):
+def connect_to_api(module, disconnect_atexit=True, return_si=False):
```
Comment (pgbidkar) [inline lib/ansible/module_utils/vmware.py:493]: I am not sure about option for return_si. But I believe that Service Instance is the object to perform any action on ManagedObjects. If it is required to provide this option then I think it should be True by default.

File: lib/ansible/module_utils/vmware.py [hunk 4]
```diff
@@ -555,6 +555,8 @@ def connect_to_api(module, disconnect_atexit=True):
     # Also removal significantly speeds up the return of the module
     if disconnect_atexit:
         atexit.register(connect.Disconnect, service_instance)
+    if return_si:
+        return service_instance, service_instance.RetrieveContent()
     return service_instance.RetrieveContent()
```
Comment (pgbidkar) [inline lib/ansible/module_utils/vmware.py:560]: I believe Service instance is right object to return after Connection to API Endpoint. Content can always be retrieved at module level.

File: lib/ansible/module_utils/vmware_spbm.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        ssl_context = client_stub.schemeArgs.get('context')
+        additional_headers = {'vcSessionCookie': session_cookie}
+        hostname = self.module.params['hostname']
+        if not hostname:
+            self.module.fail_json(msg="Please specify required parameter - hostname")
+        stub = SoapStubAdapter(host=hostname, path="/pbm/sdk", version=self.version,
+                               sslContext=ssl_context, requestContext=additional_headers)
+
+        self.spbm_si = pbm.ServiceInstance("ServiceInstance", stub)
+        self.spbm_content = self.spbm_si.PbmRetrieveServiceContent()
```
Comment (pgbidkar) [inline lib/ansible/module_utils/vmware_spbm.py:43]: Same as above comment. I feel return spbm_si here and retrieve content at the module level.

File: lib/ansible/modules/cloud/vmware/vmware_vm_storage_policy_info.py [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+                    for propertyInstance in constraint.propertyInstance:
+                        capabilities_info.append(
+                            {
+                                'id': propertyInstance.id,
+                                'value': propertyInstance.value
+                            }
+                        )
+        return capabilities_info
+
+    def ensure(self):
```
Comment (pgbidkar) [inline lib/ansible/modules/cloud/vmware/vmware_vm_storage_policy_info.py]: Method name doesnt give much info about what it does. is it possible to change?

Thanks

File: lib/ansible/module_utils/vmware.py (hunk 3 above)
Comment (Akasurde) [reply lib/ansible/module_utils/vmware.py:493]: This API exposed in module_utils and which is used by Ansible and customers who are using it in their custom modules. In order maintain backward compatibility we are setting `return_si` as `False`. So it won't affect them in any way unless they set it to True from API call.

File: lib/ansible/module_utils/vmware.py (hunk 4 above)
Comment (Akasurde) [reply lib/ansible/module_utils/vmware.py:560]: Like I said backward compatibility is maintained here.

File: lib/ansible/module_utils/vmware_spbm.py (hunk 5 above)
Comment (Akasurde) [reply lib/ansible/module_utils/vmware_spbm.py:43]: OK.

File: lib/ansible/modules/cloud/vmware/vmware_vm_storage_policy_info.py (hunk 6 above)
Comment (Akasurde) [reply lib/ansible/modules/cloud/vmware/vmware_vm_storage_policy_info.py]: OK will change to `get_policy_info` or something.

**Review by jillr — APPROVED** [review]

**Pull request merged**